components/org.wso2.carbon.identity.captcha/src/main/java/org/wso2/carbon/identity/captcha/filter/CaptchaFilter.java

```diff
@@ -67,6 +71,17 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
                 return;
             }
 
+            // Wrap Servlet request for password recovery flow as the data are in POST body of request.
+            // May need multiple reads of request body value from connectors.
+            if (servletRequest instanceof HttpServletRequest) {
+                String currentPath = ((HttpServletRequest) servletRequest).getRequestURI();
+
+                if (StringUtils.isNotBlank(currentPath) &&
+                        CaptchaUtil.isPathAvailable(currentPath, RECOVER_PASSWORD_URL)) {
+                    servletRequest = new CaptchaHttpServletRequestWrapper((HttpServletRequest) servletRequest);
```

Comment:
When calling filterChain.doFilter(servletRequest, servletResponse);  don't we need to pass the original servletRequest object. Passing wrapped CaptchaHttpServletRequestWrapper instance may cause issues in downstream logic in other filters.

Reply:
In the filter level, the post body needs to be read and therefore the request object has to be wrapped. The wrapping is done only for recovery flow through configurations. Wrapping may cause problems in authentication flow.